### PR TITLE
Pegasus: Support for HTML-only partials

### PR DESCRIPTION
**Pull request by breville**
This continues the work started in https://github.com/code-dot-org/code-dot-org/pull/42618, which allowed partials to take parameter strings which would be rendered for the user, but helps avoid the risk of a malicious translation calling into our server-side code at render time or executing any JavaScript on the client.

Now, a `{{ }}` partial that takes parameters must have `.html` source code, and at render time, we replace each `%parameter_name%` in the HTML with that parameter's sanitized value.

This PR also includes a sample of this new functionality.  The content is part of some ongoing work that will be continued in https://github.com/code-dot-org/code-dot-org/pull/42654.

Keywords: `hoc2021`

### CODE REVIEW

**Review by Hamms — COMMENTED** [review]
Unfortunately, this approach still leaves us vulnerable to HTML injection:

```diff
diff --git a/pegasus/sites.v3/code.org/public/poetry.md.partial b/pegasus/sites.v3/code.org/public/poetry.md.partial
index 09d4325f140..7b9b71ceca0 100644
--- a/pegasus/sites.v3/code.org/public/poetry.md.partial
+++ b/pegasus/sites.v3/code.org/public/poetry.md.partial
@@ -14,7 +14,7 @@ Some description goes here.
 
 ## Featured Student Creations
 
-{{ featured_project_html, title: "Coin color", author: "M", age: "13+", image: "images/csc/poetry/creation0.png", url: "/" }}
+{{ featured_project_html, title: "<script type='text/javascript'>alert('hello');</script>", author: "M", age: "13+", image: "images/csc/poetry/creation0.png", url: "/" }}
 {{ featured_project_html, title: "Real world simulation", author: "D", age: "8+", image: "images/csc/poetry/creation0.png", url: "/" }}
 {{ featured_project_html, title: "ROBOT", author: "B", age: "18+", image: "images/csc/poetry/creation0.png", url: "/" }}
 {{ featured_project_html, title: "EVADE", author: "W", age: "13+", image: "images/csc/poetry/creation0.png", url: "/" }}
```

image

**Comment by breville** [comment]
@Hamms Good catch.  I was focused on the server side but point taken.  Updated to use `ActionView` `sanitize` on parameter values which seems to mitigate this particular vulnerability.  Definitely interested if you have more thoughts on the right way to do all of this!

File: pegasus/router.rb [hunk 1]
```diff
@@ -599,8 +601,9 @@ def social_metadata
       metadata
     end
 
-    def view(uri, locals={})
+    def view(uri, locals={}, html_only=false)
       path = resolve_template('views', settings.template_extnames, uri.to_s)
+      raise "View `#{uri}` must be .html. #{File.extname(path)}" if html_only && File.extname(path) != ".html"
       raise "View '#{uri}' not found." unless path
```
Comment (Hamms) [inline pegasus/router.rb]: This leaves us vulnerable to templates which follow the common Rails naming convention of `.haml.html` I'd recommend instead doing something like:

```suggestion
      raise "View '#{uri}' not found." unless path

      if html_only
        template_is_only_html = MultipleExtnameFileUtils.all_extnames(path).all?('.html')
        raise "View `#{uri}` must be .html. #{File.extname(path)}" unless template_is_only_html
      end
```

**Review by Hamms — COMMENTED** [review]
Nice! This does still leave me feeling pretty nervous, but I'm not able to find any major vulnerabilities other than the multiple extension names one I mention below.

I would like to see a comment explaining the potential danger of this code, in particular to reduce what I see as a very high likelihood of our security precautions being evaded in the future. I'm thinking something like: 

```diff
diff --git a/pegasus/router.rb b/pegasus/router.rb
index 467e11172e9..fb8c09a146d 100644
--- a/pegasus/router.rb
+++ b/pegasus/router.rb
@@ -399,6 +399,15 @@ class Documents < Sinatra::Base
           uri = split[0]
 
           # Parse the parameters.  Adapted from https://stackoverflow.com/a/23612782.
+          #
+          # Most notably: because these parameters are being read from
+          # templates that we allow anonymous third parties to make
+          # unsupervised edits to, we need to be extremely careful to prevent
+          # code injection. Specifically, we sanitize the parameters to prevent
+          # against HTML injection and pass the `html_only` flag to the view
+          # method to ensure that we never pass parameters to templates which
+          # are capable of serverside database or filesystem access (ie,
+          # **never under any circumstances** ERB or HAML).
           locals = split[1].scan(/("(?:\\.|[^"])*"|[^\s]*):\s*("(?:\\.|[^"])*"|[^\s]*)/).
             map(&:compact).
             to_h.
```

**Comment by Hamms** [comment]
Oh, actually: I think that there is no reason at all to pass `locals` to `view` in this new implementation, since we're doing the actual interpolation manually! What do you think about something like this:

```diff
diff --git a/pegasus/router.rb b/pegasus/router.rb
index 467e11172e9..1e32e275d86 100644
--- a/pegasus/router.rb
+++ b/pegasus/router.rb
@@ -399,12 +399,27 @@ class Documents < Sinatra::Base
           uri = split[0]
 
           # Parse the parameters.  Adapted from https://stackoverflow.com/a/23612782.
+          #
+          # Most notably: because these parameters are being read from
+          # templates that we allow anonymous third parties to make
+          # unsupervised edits to, we need to be extremely careful to prevent
+          # code injection. Specifically, we sanitize the parameters to prevent
+          # against HTML injection and examine the template being rendered to
+          # ensure that we never pass parameters to templates which are capable
+          # of serverside database or filesystem access (ie, **never under any
+          # circumstances** ERB or HAML).
           locals = split[1].scan(/("(?:\\.|[^"])*"|[^\s]*):\s*("(?:\\.|[^"])*"|[^\s]*)/).
             map(&:compact).
             to_h.
             transform_values {|v| @actionview.sanitize(v.undump)}
 
-          result = view(uri, locals, !locals.empty?)
+          if locals.present?
+            path = resolve_view_template(uri)
+            template_is_only_html = MultipleExtnameFileUtils.all_extnames(path).all?('.html')
+            raise "Only HTML Partials may be passed parameters; got #{locals.inspect} for #{path.inspect}" unless template_is_only_html
+          end
+
+          result = view(uri)
           locals.each {|k, v| result.gsub!("%#{k}%", v)}
           result
         rescue
@@ -601,10 +616,14 @@ class Documents < Sinatra::Base
       metadata
     end
 
-    def view(uri, locals={}, html_only=false)
+    def resolve_view_template(uri)
       path = resolve_template('views', settings.template_extnames, uri.to_s)
-      raise "View `#{uri}` must be .html. #{File.extname(path)}" if html_only && File.extname(path) != ".html"
       raise "View '#{uri}' not found." unless path
+      return path
+    end
+
+    def view(uri, locals={})
+      path = resolve_view_template(uri)
       render_template(path, locals)
     end
```

**Comment by breville** [comment]
These suggestions have now all been incorporated, @Hamms.  Many thanks!

**Review by Hamms — APPROVED** [review]
Thank YOU for coming along with me on this whole security concerns journey! LGTM!

**Pull request merged**